### ISSUE TITLE
PM UI:  Improve package installation performance on the solution level

### ISSUE THREAD

**Issue by dtivel** [NuGet/Home#10210]
In VS 2019 16.7.6 it took ~17.5 minutes to install [System.Ben 1.0.1](https://www.nuget.org/packages/System.Ben/1.0.1) to all 82 projects in the NuGet.Client solution.
```
Time Elapsed: 00:17:27.1112177
========== Finished ==========
```
In VS 2019 16.9 P2 it took ~27 minutes to install.
```
Time Elapsed: 00:27:18.8386891
========== Finished ==========
```
Not only is 17 minutes too long to begin with, but the 10- minute regression is even worse.

I haven't investigated yet, so it's unclear exactly when the regression occurred; however, the CodeSpaces work in 16.8 is a reasonable first guess.

CC @rrelyea, @sbanni

**Comment by aortiz-msft** [comment]
cc @nkolev92

**Comment by nkolev92** [comment]
Started a thread in the triage channel, as I think the `Regression` label suggests high priority, so I want to make sure someone gets to look into this sooner rather than later.

**Comment by nkolev92** [comment]
Ideas for the initial investigation before digging deeper: 

* Does this negate the wins from https://github.com/NuGet/NuGet.Client/commit/53a7cf2b5b600cf46bc49285e1ae6f4cf887346b. 
* Is there a regression with single project installations?

**Comment by erdembayar** [comment]
@dtivel 
Hi. I can see 84 projects in Nuget.Client, so which 2 projects you're skipping when you do installation?

**Comment by erdembayar** [comment]
I just tested this on my desktop with VS 2019 Int Preview, Version 16.9.0 Preview(30701.28.main).
Time Elapsed: 00:02:25.0625671
========== Finished ==========
What kind hardware did you use? I just want to make sure we're measuring on comparable hardware.

**Comment by nkolev92** [comment]
@erdembayar 

Have you considered focusing on the relative performance differences?

In particular, use 1 machine, but run both 16.7.6 and latest 16.9.

**Comment by dtivel** [comment]
@erdembayar, I used Standard_D4s_v3.  I would not focus on using the identical configuration yet if you can reproduce relative differences easily, as @nkolev92 suggested.

When I opened this issue, 16.7.6 couldn't open dev, release-5.8.x, or release-5.7.x completely.  I think there's a deeper problem here with our branching:  we don't ensure that a release branch is rebuildable in the future by pinning the SDK version in global.json.  So, I did something hacky to get all projects to load:  I cleaned my repo (git clean -xdf), checked out release-5.7.x, ran configure.ps1, switched to release-5.6.x, and then opened that branch in both 16.7.6 and 16.8.  I don't think you need to go through these hoops to reproduce the perf issue.  It probably reproes with just 82 projects.

That said, if you want me to find exact repro steps for our buggy branching to load all 84 projects in both 16.7.6 and 16.8, I'd be happy to.  Just ping me.  I don't think it's strictly necessary though to see the perf regression.

**Comment by erdembayar** [comment]
I tested with OrchardCore(158 projects) solution. There are consistent almost `2 minutes` of regression on both OrchardCore and Nuget.Client, but it's not bad as `27 minutes`. I'll try other small., medium size solutions.

VS Release 16.7.7
1. Successfully installed '_ 1.0.0' to OrchardCore.Docs
Executing nuget actions took 1.34 sec
Time Elapsed: 00:07:25.6118670

2. Successfully installed 'System.Ben 1.0.1' to NuGet.VisualStudio.Client
Executing nuget actions took 1.7 sec
Time Elapsed: 00:00:24.1466122

VS Int Preview, Version 16.9.0  Same solutions.
1. Successfully installed '_ 1.0.0' to OrchardCore.Docs
Executing nuget actions took 1.4 sec
Time Elapsed: 00:09:20.5624785

2. NuGet.VisualStudio.Client
Time Elapsed: 00:02:25.0625671

**Comment by erdembayar** [comment]
I just followed @dtivel 's regression reproduce guide. 
But was unable to reproduce it on my local, they were almost identical. Repository used for this test was : NugetClient "release-5.6.x" branch
I shared my logs to Damon on Teams and I am co-assigning bug to him. He'll update the repro steps.

**VS Release 16.7.7** (VSTelemetryMonitor was open that could be contributed to 1 min slower result here.)
_Successfully installed 'System.Ben 1.0.1' to NuGet.VisualStudio.Client
Executing nuget actions took 1.21 sec
Time Elapsed: 00:08:06.4404173_
========== Finished ==========

**VS Int Preview, Version 16.9.0**  (VSTelemetryMonitor closed)

_Successfully installed 'System.Ben 1.0.1' to NuGet.VisualStudio.Client
Executing nuget actions took 6.56 sec
Time Elapsed: 00:07:02.1688446_
========== Finished ==========

**Comment by erdembayar** [comment]
@dtivel 
Is this issue still valid?

**Comment by dtivel** [comment]
I think so.  I haven't retested recently, but nothing changed in this area.

**Comment by erdembayar** [comment]
@aortiz-msft 
FYI : Here is repro steps from Damon.
Connect to test Azure machine.

1.  Open PowerShell and path to F:\git\NuGet.Client.Test.
2.  In PowerShell execute:  

git clean -xdf
.\configure.ps1

3.  Launch Visual Studio Installer.
4.  Launch Visual Studio Enterprise 2019 (16.7.7).
5.  Open F:\git\NuGet.Client.Test\NuGet.sln.
6.  In Solution Explorer right-click on the solution node and select "Manage NuGet Packages for Solution..."
7.  In PM UI navigate to the Browse tab.
8.  Search for System.Ben.
9.  In the left-hand list pane select the System.Ben package.
10.  In the right-hand details pane:
    a.  Select all projects.
    b.  Expand the Options section and disable (uncheck) "Show preview window".
    c.  Important: Wait until the solution has fully loaded and system-wide CPU and disk usage drops to ~0% (per Task Manager).
    d.  Click the Install button.
11.  When install completes open the Output tool window's "Package Manager" pane and check the elapsed time.  Note:  there may be multiple elapsed times.  You can either use the largest elapsed time or sum all elapsed times.
12.  Close Visual Studio.
13.  Repeat steps 2-12 but in step 4 launch "Visual Studio Enterprise 2019 Int Preview" instead.

Thanks,

Damon

**Comment by nkolev92** [comment]
Fixed in 16.10 by partners, internal [issue link](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1281052)

**Comment by nkolev92** [comment]
Tweaked the title to make it release note happy.
